src/cirq_iqm/iqm_client.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2020–2021 Cirq on IQM developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Implements a client for calling the IQM backend
+"""
+import time
+from dataclasses import dataclass
+import json
+import requests
+from enum import Enum
+from datetime import datetime
+
+TIMEOUT_SECONDS = 10
+SECONDS_BETWEEN_CALLS = 1
+
+
+class IQMException(Exception):
+    pass
+
+
+class ApiTimeoutError(IQMException):
+    pass
+
+
+class RunStatus(str, Enum):
+    PENDING = "pending"
+    READY = "ready"
+    FAILED = "failed"
+
+
+@dataclass(frozen=True)
+class IQMInstruction:
+    name: str
+    qubits: list[str]
+    args: dict
+
+
+@dataclass(frozen=True)
+class IQMCircuit:
+    name: str
+    args: dict
+    instructions: list[IQMInstruction]
+
+
+@dataclass(frozen=True)
+class QubitMapping:
+    logical_name: str
+    physical_name: str
+
+
+@dataclass(frozen=True)
+class RunResult():
+    status: RunStatus
+    measurements: dict[str:list[list]] = None
+    message: str = None
+
+    @staticmethod
+    def parse(input: dict):
+        input_copy = input.copy()
+        return RunResult(status=RunStatus(input_copy.pop("status")), **input_copy)
+
+
+class IQMBackendClient:
+    def __init__(self, url: str, token: str):
+        self._token = token
+        self._base_url = url
+
+    def submit_circuit(self, circuit: IQMCircuit, mappings: list[QubitMapping] = {}, shots: int = 1) -> int:
+        """
+        Submits circuit to the IQM backend
+        Args:
+            circuit: Circuit to be executed on the IQM backend
+            mappings: Mappings of human-readable names to physical names
```

Comment:
I would call the parameter `qubit_mapping`, since it defines a single mapping.

Reply:
It's a list of QubitMapping instances
If we want to make call it mapping, we need to rename QubitMapping to something else first

Reply:
Let's rename QubitMapping to SingleQubitMapping then.

Reply:
Renamed